### PR TITLE
Support for computed goto

### LINKED ISSUE #864: C Dispatch Tables
GCC supports using labels as values in order to implement dispatch tables: https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html in the interest of a slight performance improvement over using a switch statement.

Does CC65 have any way to do this or similar? Or is it only possible in ASM?

### PR DESCRIPTION
**Pull request by clbr**
Depends on #873, and includes those commits so that the autobuilder will succeed.
Closes #864.

### CODE REVIEW

File: src/cc65/codeseg.c [hunk 1]
```diff
@@ -426,8 +426,10 @@ static CodeEntry* ParseInsn (CodeSeg* S, LineInfo* LI, const char* L)
         unsigned Hash = HashStr (Arg) % CS_LABEL_HASH_SIZE;
         Label = CS_FindLabel (S, Arg, Hash);
 
-        /* If we don't have the label, it's a forward ref - create it */
-        if (Label == 0) {
+        /* If we don't have the label, it's a forward ref - create it unless
+        ** it's an external function.
+        */
+        if (Label == 0 && (OPC->OPC != OP65_JMP || IsLocalLabelName(Arg)) ) {
```
Comment (oliverschmidt) [inline src/cc65/codeseg.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/expr.c [hunk 2]
```diff
@@ -696,6 +696,22 @@ static void Primary (ExprDesc* E)
...
 
+        case TOK_BOOL_AND:
+            /* A computed goto label address */
+            if (IS_Get (&Standard) >= STD_CC65) {
+                NextToken ();
+                SymEntry* Entry = AddLabelSym (CurTok.Ident, SC_REF | SC_GOTO_IND);
+                /* output its label */
+                E->Flags = E_RTYPE_RVAL | E_LOC_STATIC;
+                E->Name = Entry->V.L.Label;
+                E->Type = PointerTo(type_void);
```
Comment (oliverschmidt) [inline src/cc65/expr.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/goto.c [hunk 3]
```diff
@@ -54,21 +62,97 @@ void GotoStatement (void)
...
+        NextToken ();
+
+        /* arr[foo], we only support simple foo for now */
+        if (CurTok.Tok == TOK_IDENT &&
+            (arr = FindSym (CurTok.Ident))) {
+            NextToken ();
+            ConsumeLBrack ();
+
+            /* Find array size */
+            if (!IsTypeArray(arr->Type) || SizeOf(arr->Type) == 0 ||
```
Comment (oliverschmidt) [inline src/cc65/goto.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/goto.c [hunk 4]
```diff
@@ -54,21 +62,97 @@ void GotoStatement (void)
...
+                SizeOf(GetElementType(arr->Type)) != 2)
```
Comment (oliverschmidt) [inline src/cc65/goto.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/goto.c [hunk 5]
```diff
@@ -54,21 +62,97 @@ void GotoStatement (void)
...
+                Error ("Expected array");
+            if (GetElementCount(arr->Type) > 127)
```
Comment (oliverschmidt) [inline src/cc65/goto.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/goto.c [hunk 6]
```diff
@@ -54,21 +62,97 @@ void GotoStatement (void)
...
+                Error ("Only arrays with <= 127 labels are supported, got %lu",
+                       GetElementCount(arr->Type));
```
Comment (oliverschmidt) [inline src/cc65/goto.c]: Please follow the predominant code style (space before parameter list).

File: src/cc65/symtab.c [hunk 7]
```diff
@@ -739,21 +739,23 @@ SymEntry* AddLabelSym (const char* Name, unsigned Flags)
...
-
-                /* Optimizer will need the information about the value of SP adjustment
-                ** later, so let's preserve it.
-                */
-                E = NewSymEntry (LocalLabelName (DOR->LateSP_Label), SC_SPADJUSTMENT);
-                E->V.SPAdjustment = StackPtr - DOR->StackPtr;
-                AddSymEntry (SPAdjustTab, E);
+                if (DOR->Flags & SC_GOTO) {
+                    SymEntry *E;
+                    g_userodata();
```
Comment (oliverschmidt) [inline src/cc65/symtab.c]: Please follow the predominant code style (space before parameter list).

**Comment by clbr** [comment]
Will fix the missing spaces, thanks.

As mentioned in #864, it allows efficient jump tables from C. A switch-case or a cascade of ifs can take 30x or more cpu vs what a jump table jump takes, and it also changes the timing from variable to constant. As a plus, it's also smaller than the alternatives.

File: test/val/computedgoto.c [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	op5:
+	val2 += 7;
+	return;
+
+	op6:
+	val2 += 9;
+	return;
+}
+
+int main() {
```
Comment (greg-king5) [inline test/val/computedgoto.c]: Please make the parameter list `void`.

File: test/val/jmp-callax.c [hunk 9]
```diff
@@ -0,0 +1,21 @@
+static unsigned char val;
+
+static void foo(void) {
+	val = 5;
+}
+
+static void wrap() {
```
Comment (greg-king5) [inline test/val/jmp-callax.c]: Please make the parameter list `void`.

File: test/val/jmp-callax.c [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+	asm("lda #<%v", foo);
+	asm("ldx #>%v", foo);
+	asm("jmp callax");
+
+}
+
+int main() {
```
Comment (greg-king5) [inline test/val/jmp-callax.c]: Parameter list `void`, please.

**Comment by clbr** [comment]
> That test shouldn't be done if the previous test says that the variable isn't an array.

Isn't that what the Error does? Stops execution since we can't compile this.

**Comment by greg-king5** [comment]
> > That test shouldn't be done if the previous test says that the variable isn't an array.
> 
> Isn't that what `Error()` does:  stops execution since we can't compile this?

`Warning()` and `Error()` show diagnostics on stderr, and count them.  Compilation continues, so that other problems can be found.  At the end, the output file is written only if no errors were counted.

**Comment by clbr** [comment]
I checked the source; it's only fatal after a few errors, my mistake, I remembered it was fatal instantly.

File: src/cc65/goto.c [hunk 11]
```diff
@@ -54,21 +62,102 @@ void GotoStatement (void)
...
+
+        /* arr[foo], we only support simple foo for now */
+        if (CurTok.Tok == TOK_IDENT &&
+            (arr = FindSym (CurTok.Ident))) {
+            NextToken ();
+
+            /* Find array size */
+            if (!IsTypeArray (arr->Type) || SizeOf (arr->Type) == 0 ||
+                SizeOf (GetElementType(arr->Type)) != 2) {
+                Error ("Expected array");
```
Comment (greg-king5) [inline src/cc65/goto.c]: The current implementation requires static label arrays.  You should add a test for the `SC_STATIC` flag.  And, make the error message say, `"Expected a static array"`.

File: src/cc65/codeseg.c [hunk 12]
```diff
@@ -426,8 +426,10 @@ static CodeEntry* ParseInsn (CodeSeg* S, LineInfo* LI, const char* L)
         unsigned Hash = HashStr (Arg) % CS_LABEL_HASH_SIZE;
         Label = CS_FindLabel (S, Arg, Hash);
 
-        /* If we don't have the label, it's a forward ref - create it */
-        if (Label == 0) {
+        /* If we don't have the label, it's a forward ref - create it unless
+        ** it's an external function.
+        */
+        if (Label == 0 && (OPC->OPC != OP65_JMP || IsLocalLabelName (Arg)) ) {
```
Comment (greg-king5) [inline src/cc65/codeseg.c]: I'm confused by commit dafac1d4bd4663dc4fa9b35f55e5413100c7e22f.  The commit message says, "Only for jumps".  But, that opcode test creates labels for anything _except_ jumps!  Is that test wrong; or, do I misunderstand the message?

**Comment by clbr** [comment]
About the "only for jumps" message, expanding it slightly:
"Do the test only for jumps, because the lib code has manual asm statements like asm("bne L1"). These L1-named labels don't match the IsLocalLabel function, and not emitting that label would break the function".

**Comment by polluks** [comment]
Maybe a hint in `cc65.html#s5`?

**Comment by oliverschmidt** [comment]
> [...] and I don't understand it's importance.

Now with the docs addition I get it...

@greg-king5: Thanks for your review! As far as I can see this pr is ready to be merged from your pov, right? In that case please either acknowledge or merge right away.

@polluks: Thanks for the hint on the docs.

**Comment by greg-king5** [comment]
> As mentioned in #864, it allows efficient jump tables from C. A `switch-case` or a cascade of `ifs` can take 30x or more CPU time vs what a jump-table jump takes.  And, it also changes the timing from variable to constant. As a plus, it's also smaller than the alternatives.

Those advantages are useful especially when making game animations.

Merging ...

**Pull request merged**